contracts/atomic-order-example/src/contract.rs

```diff
@@ -0,0 +1,188 @@
+    let deposit_message = SubMsg::new(create_deposit_msg(
+        contract.to_string(),
+        subaccount_id,
+        coins.clone(),
+    ));
+    let order_message = SubMsg::reply_on_success(
+        create_spot_market_order_msg(contract.into_string(), order),
+        ATOMIC_ORDER_REPLY_ID,
+    );
+    let response = Response::new()
+        .add_submessage(deposit_message)
+        .add_submessage(order_message);
+
+    let cache = SwapCacheState { sender_address: info.sender.to_string(), deposited_amount: coins.clone() };
+    CACHE.save(deps.storage, &cache)?;
+
+    return Ok(response);
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(
+    deps: DepsMut,
+    _env: Env,
+    msg: Reply,
+) -> Result<Response<InjectiveMsgWrapper>, StdError> {
+    // deps.api.debug(format!("SC ----> Received reply (data): {}", msg.clone().result.unwrap().data.unwrap()).as_str(), );
+    // let events = msg.clone().result.unwrap().events;
+    // deps.api.debug(format!("SC ----> Received reply (events count): {}", events.len()).as_str());
+    match msg.id {
+        ATOMIC_ORDER_REPLY_ID => handle_atomic_order_reply(deps, msg),
+        id => Err(StdError::generic_err(format!("Unknown reply id: {}", id))),
+    }
+}
+
+fn handle_atomic_order_reply(
+    deps: DepsMut,
+    msg: Reply,
+) -> Result<Response<InjectiveMsgWrapper>, StdError> {
+    let dec_scale_factor: FPDecimal = FPDecimal::from(1000000000000000000 as i128);
+    let mut data = msg.result.unwrap().data.unwrap().to_vec();
+    let _ = parse_protobuf_string(&mut data, 1); // order hash - we need to read to advance pointer
+
+    let trade_result = parse_protobuf_bytes(&mut data, 2).unwrap();
+    let mut trade_data = trade_result.unwrap().to_vec();
+    let field1 = parse_protobuf_string(&mut trade_data, 1).unwrap();
+    let field2 = parse_protobuf_string(&mut trade_data, 2).unwrap();
+    let field3 = parse_protobuf_string(&mut trade_data, 3).unwrap();
+    let quantity = FPDecimal::from_str(&field1)? / dec_scale_factor;
+    let price = FPDecimal::from_str(&field2)? / dec_scale_factor;
+    let fee = FPDecimal::from_str(&field3)? / dec_scale_factor;
+
+    let config = STATE.load(deps.storage)?;
+    let cache = CACHE.load(deps.storage)?;
+
+    let recipient_subaccount_id = default_subaccount_id(&Addr::unchecked(&cache.sender_address));
+
+    let msg1 = create_external_transfer_msg(
+        config.owner.clone().into_string(),
+        default_subaccount_id(&config.owner),
+        recipient_subaccount_id.clone(),
+        Coin::new(u128::from(quantity), config.base_denom.clone()),
+    );
+
+    let paid = quantity * price + fee;
+    let to_send = cache.deposited_amount.amount - Uint128::from(u128::from(paid));
+    let msg2 = create_external_transfer_msg(
+        config.owner.clone().into_string(),
+        default_subaccount_id(&config.owner),
+        recipient_subaccount_id,
+        Coin::new(u128::from(to_send), config.quote_denom),
+    );
```

Comment:
Maybe better to send it to the bank balances back where the funds originally came from. So still two messages required, but those ones:

1. MsgWithdraw
2. Bank Balances Transfer